openvdb/points/AttributeArrayString.h

```diff
@@ -49,27 +49,68 @@ struct StringCodec
 };
 
 
-using StringAttributeArray = TypedAttributeArray<StringIndexType, StringCodec<false>>;
+using StringAttributeArray = TypedAttributeArray<Index, StringCodec<false>>;
 
 
 ////////////////////////////////////////
 
 
+/// Class to compute a string->index map from all string:N metadata
+class OPENVDB_API StringMetaCache
+{
+public:
+    using UniquePtr = std::unique_ptr<StringMetaCache>;
+    using ValueMap = std::unordered_map<Name, Index>;
+
+    StringMetaCache() = default;
+    explicit StringMetaCache(const MetaMap& metadata);
+
+    /// Return @c true if no string elements in metadata
+    bool empty() const { return mCache.empty(); }
+    /// Returns the number of string elements in metadata
+    size_t size() const { return mCache.size(); }
+
+    /// Clears and re-populates the cache
+    void reset(const MetaMap& metadata);
+
+    /// Insert a new element in the cache
+    void insert(const Name& key, Index index);
+
+    /// Retrieve the value map (string -> index)
+    const ValueMap& map() const { return mCache; }
+
+private:
+    ValueMap mCache;
+}; // StringMetaCache
+
+
+////////////////////////////////////////
+
+
+/// Class to help with insertion of keyed string values into metadata
 class OPENVDB_API StringMetaInserter
 {
 public:
-    StringMetaInserter(MetaMap& metadata);
+    explicit StringMetaInserter(MetaMap& metadata);
+
+    /// Returns @c true if key exists
+    bool hasKey(const Name& key) const;
+    /// Returns @c true if index exists
+    bool hasIndex(Index index) const;
 
-    /// Insert the string into the metadata
-    void insert(const Name& name);
+    /// Insert the string into the metadata, hint is used if non-zero
+    /// Returns the chosen index which will match hint if not in use
+    Index insert(const Name& name, Index hint = Index(0));
```

Comment:
I think we should try to get in the habbit of proper doxygen with any function changes

Reply:
And perhaps a bit more of a description about the intended usage of hint and how it's expected to map the metadata name index

Reply:
I'll add a bit more description about the hint. I'm not going to fix all the doxygen for this class though as it's not one that I expect users to interact with directly. I don't mind if you want to make a "good first issue" in our GitHub issues for fixing some of these up though.

Reply:
I don't agree with that mentality. New users should not be writing doxygen, especially at this level - that's our job. We should get in the habit of writing good documentation whenever we introduce a new method as we're most likely best situated to explain it. And in regards to "not one that I expect users to interact with directly" - I'm a user and it wasn't clear to me without looking at the implementation! I wasn't suggesting you you update this entire class now, but IMO we should keep on top of doxygen with new methods.

Reply:
Sorry, that's not what I meant. If the comments in the code are not sufficiently clear, that's totally valid (and appreciated) feedback. I just mean that I don't agree with a policy that every piece of documentation should be formatted for Doxygen. For important public API, sure. However, parameter names still show up in Doxygen, so provided that the description is clear and unambiguous, I don't think it should be a strict requirement to ensure correct use of `@param` for instance as I don't think that adds value to the project.

Reply:
Okay that makes sense to me now